### ISSUE TITLE
AQL performance v3.2.x vs v3.1.1

### ISSUE THREAD

**Issue by VladoDemcak** [arangodb/arangodb#3917]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [x] 3.2

Mode:
- [x] Single-Server

Storage-Engine:
- [x] mmfiles

On this operating system:
- [x] Linux 
  - [x] docker - official docker library

### this is an AQL-related issue:
- [x] I'm using graph features

I'm issuing AQL via:
- [x] web interface with this browser:     running on this OS:
- [x] this Driver: arango-java-driver

I've run `db._explain("<my aql query>")` and it didn't shed more light on this.
The AQL query in question is:  AQL performance in v3.2.x docker takes too long when we have higher depth level. In version 3.1.1 it worked fine with same values. Is it expected behavior? 

##  These are the steps to reproduce:
1) open the browser on http://127.0.0.1:8529
2) use database with simple graph which is relevant to query in step 5
3) create simple graph structure with 10 nodes 
4) Select Queries
5) execute AQL
```
FOR vertex, e, path IN 0..2147483647 INBOUND @rootVertex GRAPH 'graph' FILTER LAST(path.vertices)._key IN @ids RETURN vertex`
// this takes **~25 seconds**
```
and this with depth lesser than the previous one:
```
FOR vertex, e, path IN 0..7483647 INBOUND @rootVertex GRAPH 'graph' RETURN vertex
// takes only **0.132970 seconds**
```

This happens with arango docker `3.2.X` version **but worked** fine with `3.1.1` (without huge delay)
...

**Comment by dothebart** [comment]
Hi, 
can you post the `explain()` output of both versions?

**Comment by VladoDemcak** [comment]
Hi,
thanks for quick response. Please check explain below:


with depth 7483647 
```
Query string:
 FOR vertex, e, path IN 0..7483647 INBOUND @rootVertex GRAPH @graph RETURN vertex

Execution plan:
 Id   NodeType        Est.   Comment
  1   SingletonNode      1   * ROOT
  2   TraversalNode      0     - FOR vertex  /* vertex */ IN 0..7483647  /* min..maxPathDepth */ INBOUND 'vertices-cccccccccccccccccccc/0001045E00040002-1303-0030319D' /* startnode */  GRAPH 'graph'
  3   ReturnNode         0     - RETURN vertex

Indexes used:
 By   Type   Collection                 Unique   Sparse   Selectivity   Fields               Ranges
  2   edge   edges-iiiiii               false    false       100.00 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-ddddddddddddddd      false    false       100.00 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-ggggggggggggg        false    false        75.00 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-ffffffffffffffffff   false    false        68.75 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-eeeeeeeeeeeeeee      false    false       100.00 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-hhhhhhhhhhhhhhhh     false    false        83.33 %   [ `_from`, `_to` ]   base INBOUND

Traversals on graphs:
 Id   Depth        Vertex collections                                                Edge collections                                                                                                                    Options                                   Filter conditions
  2   0..7483647   vertices-aaaaaa, vertices-bbbbbb, vertices-cccccccccccccccccccc   edges-ddddddddddddddd, edges-eeeeeeeeeeeeeee, edges-ffffffffffffffffff, edges-ggggggggggggg, edges-hhhhhhhhhhhhhhhh, edges-iiiiii   uniqueVertices: none, uniqueEdges: path   

Optimization rules applied:
 Id   RuleName
  1   optimize-traversals

  
```


with depth 2147483647 
```  
Query string:
 FOR vertex, e, path IN 0..2147483647 INBOUND @rootVertex GRAPH @graph RETURN vertex

Execution plan:
 Id   NodeType        Est.   Comment
  1   SingletonNode      1   * ROOT
  2   TraversalNode      0     - FOR vertex  /* vertex */ IN 0..2147483647  /* min..maxPathDepth */ INBOUND 'vertices-cccccccccccccccccccc/0001045E00040002-1303-0030319D' /* startnode */  GRAPH 'graph'
  3   ReturnNode         0     - RETURN vertex

Indexes used:
 By   Type   Collection                 Unique   Sparse   Selectivity   Fields               Ranges
  2   edge   edges-iiiiii               false    false       100.00 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-ddddddddddddddd      false    false       100.00 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-ggggggggggggg        false    false        75.00 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-ffffffffffffffffff   false    false        68.75 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-eeeeeeeeeeeeeee      false    false       100.00 %   [ `_from`, `_to` ]   base INBOUND
  2   edge   edges-hhhhhhhhhhhhhhhh     false    false        83.33 %   [ `_from`, `_to` ]   base INBOUND

Traversals on graphs:
 Id   Depth           Vertex collections                                                Edge collections                                                                                                                    Options                                   Filter conditions
  2   0..2147483647   vertices-aaaaaa, vertices-bbbbbb, vertices-cccccccccccccccccccc   edges-ddddddddddddddd, edges-eeeeeeeeeeeeeee, edges-ffffffffffffffffff, edges-ggggggggggggg, edges-hhhhhhhhhhhhhhhh, edges-iiiiii   uniqueVertices: none, uniqueEdges: path   

Optimization rules applied:
 Id   RuleName
  1   optimize-traversals
```

**Comment by dothebart** [comment]
Ok, the optimizer has the problem here, as we can already see by your `explain` time; We have been able to reproduce this.

**Comment by mchacki** [comment]
Thanks for spotting and reporting it.
Fixes are available on 3.2 onwards and will be part of next releases.
Does not apply to 3.1 and earlier
